daemon/migration.go

```diff
@@ -0,0 +1,143 @@
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/log"
+	"github.com/containerd/platforms"
+	"github.com/docker/docker/api/types/backend"
+	"github.com/docker/docker/container"
+	"github.com/docker/docker/image"
+	"github.com/docker/docker/internal/multierror"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+func migrateContainerOS(ctx context.Context,
+	migration platformReader,
+	ctr *container.Container,
+) {
+	deduced, err := deduceContainerPlatform(ctx, migration, ctr)
+	if err != nil {
+		log.G(ctx).WithFields(log.Fields{
+			"container": ctr.ID,
+			"error":     err,
+		}).Warn("failed to deduce the container architecture")
+		ctr.ImagePlatform.OS = ctr.OS //nolint:staticcheck // ignore SA1019
+		return
+	}
+
+	ctr.ImagePlatform = deduced
+}
+
+type platformReader interface {
+	ReadPlatformFromConfigByImageManifest(ctx context.Context, desc ocispec.Descriptor) (ocispec.Platform, error)
+	ReadPlatformFromImage(ctx context.Context, id image.ID) (ocispec.Platform, error)
+}
+
+// deduceContainerPlatform tries to deduce `ctr`'s platform.
+// If both `ctr.OS` and `ctr.ImageManifest` are empty, it's not possible to deduce
+// the container platform and an error is returned.
+// Otherwise:
+// - `ctr.ImageManifest.Platform` is used, if it exists and is not empty.
+// - The platform from the manifest's config is used, if `ctr.ImageManifest` exists
+// and we're able to load it's config from the content store.
+// - The platform found by loading the image from the image service by ID (using
+// `ctr.ImageID`) is used – this looks for the best *present* matching manifest in
+// the store.
+func deduceContainerPlatform(
+	ctx context.Context,
+	migration platformReader,
+	ctr *container.Container,
+) (ocispec.Platform, error) {
+	if ctr.OS == "" && ctr.ImageManifest == nil { //nolint:staticcheck // ignore SA1019 because we are testing deprecated field migration
+		return platforms.DefaultSpec(), nil
+	}
+
+	var errs []error
+
+	isValidPlatform := func(p ocispec.Platform) bool {
+		return p.OS != "" && p.Architecture != ""
```

Comment:
Silly question; should we consider only an `OS` be "valid"?

Reply:
I don't think it should be valid - the `architecture` field is required by the OCI spec:

> - **architecture** _string_, REQUIRED
>
>  The CPU architecture which the binaries in this image are built to run on.
>  Configurations SHOULD use, and implementations SHOULD understand, values listed in the Go Language document for [`GOARCH`][go-environment].
> (https://github.com/opencontainers/image-spec/blob/main/config.md#properties)